crates/codegen/parser/src/rust_lib_code_generator.rs

```diff
@@ -0,0 +1,156 @@
+use quote::quote;
+use std::path::PathBuf;
+
+use codegen_utils::context::CodegenContext;
+
+use super::code_generator::CodeGenerator;
+
+impl CodeGenerator {
+    pub fn write_rust_lib_sources(&self, codegen: &mut CodegenContext, output_dir: &PathBuf) {
+        self.write_common_sources(codegen, output_dir);
+
+        codegen
+            .write_file(
+                &output_dir.join("mod.rs"),
+                &quote! {
+                   pub mod kinds;
+                   pub mod cst;
+                   pub mod cst_visitor;
+                   pub mod language;
+                   pub mod parser_output;
+                   mod scanners;
+                   mod parsers;
+                }
+                .to_string(),
+            )
+            .unwrap();
+
+        codegen
+            .write_file(
+                &output_dir.join("cst.rs"),
+                include_str!("../../parser_templates/src/shared/cst.rs"),
+            )
+            .unwrap();
+
+        codegen
+            .write_file(
+                &output_dir.join("cst_visitor.rs"),
```

Comment:
For loading these files, I suggest using `CodeGenContext::read_file()`. Because it will do dependency tracking correctly, and without it, Cargo incremental builds/checks will not work correctly if these files are updated.

Reply:
It also enables calculating paths at runtime. Using paths relative to repo root, for example `codegen.repo_root.join("crates/solidity/inputs/schema/grammar/manifest.yml")`, makes refactoring/moving these files around much easier, as I can just do a string search for their full paths (or one of the parent dirs) across the repo to find all usages. But your workflow might be different.

Reply:
Can't - these are read at compile time.

Reply:
But does it have to? i.e.

```rust
let cst_file = codegen.read_file(codegen.repo_root.join("....../parser_templates/src/shared/cst.rs"));
codegen.write_file(&output_dir.join("cst.rs"), cst_file);
```

We can even add a common `codegen.copy_file(source, destination)` that calls both if we have a lot of these 😄 

This means that:

1. We don't need to inject the template source in the generated binary.
2. We don't need to recompile the code generator if the template changes. This is important when editing the template manually, as it makes the output ready instantly.
3. But most importantly, print the correct cargo directives for them, similar to any other codegen input.

Reply:
I'm not keen on this because it means the binary can never be standalone.

Reply:
F2F: we decided to unify the FS API we use for now, and if/when we need to run stand-alone, we can probably replace the underlying `std::fs::read_file()` call with something like this:

```rust
#[cfg(IS_SLANG_STANDALONE_BUILD)]
return include_str!(path);

#[cfg(not(IS_SLANG_STANDALONE_BUILD))]
return std::fs::read_file(path);
```

Reply:
done